### ISSUE TITLE
Refactor Optimizer Cost Model to Return PlanNode Cardinalities

### ISSUE THREAD

**Issue by apavlo** [cmu-db/terrier#703]
@malin1993ml needs the cardinality estimates for plan nodes for his self-driving models. Right now the `AbstractCostModel` API only produces a single double as the cost. Its internal cardinality estimates are not exposed outside of the cost model:

https://github.com/cmu-db/terrier/blob/master/src/include/optimizer/cost_model/abstract_cost_model.h#L28

We need to modify this API to return a mapping from plan nodes to their cardinality estimates in addition to the cost. This map should be based on internal plan node identifiers. We **cannot** store the cardinality estimate in the plan nodes because we need to be able to reuse them across multiple invocations.

1. Create a new strong typedef `plan_node_id_t` in [optimizer_defs.h](https://github.com/cmu-db/terrier/blob/master/src/include/optimizer/optimizer_defs.h). Thee ids are unique within a query plan but not across the entire system (e.g., two plan nodes in different queries can have the same id).

2. Modify `BaseOperatorNode` to add a new member `plan_node_id_`. This should be a mandatory value passed in to the constructor. Make sure 

3. Modify `AbstractPlanNode` to also add `plan_node_id_`. Again, this should be a mandatory value passed in to the object's constructor and its builder. Make sure to also serialize/deserialize the id in `ToJson`/`FromJson`. I don't think that we should include the `plan_node_id_` in `AbstractPlanNode`'s equals and hash methods, but we should check with @17zhangw / @esargent28 .

4. Create a struct called `CostModelEstimate` in [abstract_cost_model.h](https://github.com/cmu-db/terrier/blob/master/src/include/optimizer/cost_model/abstract_cost_model.h). This should contain a `cost` (`double`) and `cardinalities` (`unordered_map<plan_node_id_t ,double>`).
    **TODO:** We should decide whether the cost mode creates the cardinalities map each time it is invoked or whether we can pass in the map by reference so that we can reuse the data structure.

5. Modify the optimizer to use a simple counter to generate new `plan_node_id_t` for each `BaseOperatorNode` that it generates.

6. We then need to modify `PlanGenerator` to support passing the `BaseOperatorNode` id to their corresponding `AbstractPlanNode`. This will ensure that we can reuse the cardinality estimates in the final query plan without re-generating the map.

**Comment by xinzhu-cai** [comment]
https://github.com/cmu-db/terrier/pull/759

**Comment by malin1993ml** [comment]
We had a discussion with @apavlo and @17zhangw to separate the logic of getting the cardinality estimation numbers from the cost models.

After looking through the code, I think the logic to derive the cardinality (and the number of rows) estimations are already in the `StatsCalculator`. The estimation for the number of rows is set for each `Group` though the `Group::SetNumRows()` function, like [here](https://github.com/cmu-db/terrier/blob/master/src/include/optimizer/group.h#L124). (P.S. we need to extend that to record the cardinality). I think the most straightforward way to expose those stats numbers is in fact to populate the map from `plan_node_id_t` to a `PlanStats` struct when generating the query plan with `PlanGenerator`. The stats for the operators that are not in the final plan tree can cause additional storage overhead without much value. And all these operators are blown away anyway after the optimization finished.

@apavlo @17zhangw What do you guys think?

**Comment by 17zhangw** [comment]
I agree with the approach Lin described above. There already exists code that sets the number of rows (which can be easily extended to add cardinality). Then generating the `PlanStats` mapping can be done in a single pass through the `PlanGenerator`.

Under this case, I don’t think logical/physical operators require `plan_node_id_t` anymore?

**Comment by malin1993ml** [comment]
Thanks for the input, @17zhangw. Strictly speaking we may not need the `plan_node_id_t` in the operators, but I think it doesn't really hurt to keep it here. I think it's generally a nice thing to have identifiers for the operators (with very little overhead) and we may need to use them in the future.
